File: test/test_script/unit_test_iec60730_bist.py
#!/usr/bin/env python
# License
# <b>Copyright 2021 Silicon Laboratories Inc. www.silabs.com</b>
# *******************************************************************************
#
# The licensor of this software is Silicon Laboratories Inc. Your use of this
# software is governed by the terms of Silicon Labs Master Software License
# Agreement (MSLA) available at
# www.silabs.com/about-us/legal/master-software-license-agreement. This
# software is distributed to you in Source Code format and is governed by the
# sections of the MSLA applicable to Source Code.
# *******************************************************************************

from binascii import hexlify
import random
import os
import sys
from time import sleep

from iec60730_get_report_unit_test import *
from iec60730_test_base import *
import jlink, jlink_constants
import unittest
import enums

def int_to_bytes(number: int) -> bytes:
    return number.to_bytes(length=(8 + (number + (number < 0)).bit_length()) // 8, byteorder='big', signed=True)


##  IEC60730 get unit test result
#
# @details This class runs IEC60730 get unit test results
#
# Hardware setup:
#   Connect DUT to PC. No additional hardware setup required.

class iec60730_unit_test_result(unittest.TestCase, iec60730TestBase):

  def get_unit_test_result(self):
    def get_result(self):
      self.adapter.halt()
      self.adapter.clear_all_breakpoints()

      self.adapter.set_breakpoint(address = asm_labels['IEC60730_UNIT_TEST_END'],
                                  typeflags=enums.JLinkBreakpoint.ANY)

      self.adapter.reset()
      pc = self.adapter.run_to_breakpoint(3)
      logging.info("Halted at label: " + self.get_label(pc))
      self.assertEqual(pc, asm_labels['IEC60730_UNIT_TEST_END'],\
                      "DUT failed to stop at IEC60730_UNIT_TEST_END" \
                      " , pc = " + hex(pc))
      #logging.info("Halted at label: " + self.get_label(pc))
      current_test_numth = variables['current_test_numth']
      self.assertNotEqual(0, current_test_numth,\
                        "Can not read current_test_numth address")
      reg = self.adapter.memory_read8(current_test_numth, 1, zone=None)

      current_test_numth = variables['current_test_numth']
      self.assertNotEqual(0, current_test_numth,\
                      "Can not read current_test_numth address")
      current_test_numth = self.adapter.memory_read8(current_test_numth, 1, zone=None)

      test_case_result = variables['test_case_result']
      self.assertNotEqual(0, test_case_result,\
                     "Can not read test_case_result address")
      result = self.adapter.memory_read(test_case_result,current_test_numth[0])
      return result
    def connect_device(self):
      self.pre_exec_test()
      logging.info("----------------- Start get unit test result")
      result = get_result(self)
      logging.info("Get unit test result: Done. See unit_test_report.log")
      logging.info("----------------- End get unit test result")
      self.post_exec_test()
      file_test_case = "../unit_test/src/unit_test_iec60730_bist.c"
      log_report = '../../log/unit_test_iec60730_bist.log'
      get_report_unit_test(result,file_test_case,log_report)
    return connect_device(self)

if __name__ == "__main__":

  chip_name = os.getenv('CHIP')

  app_type = os.getenv('APP_TYPE')

  lst_file_path = os.getenv('LST_PATH')

  adapter_serial_no = os.getenv('ADAPTER_SN')

  lib_path = os.getenv('JLINK_PATH')

  while len(sys.argv) > 1:
    line = sys.argv.pop()
    if len(sys.argv) == 1:
      compiler = line

  print("Compiler: "+compiler)
  
  iec60730_unit_test_result = iec60730_unit_test_result()
  iec60730_unit_test_result.env_setup(adapter_serial_no,
                                    chip_name,
                                    lst_file_path,
                                    lib_path,
                                    app_type,
                                    compiler)
  iec60730_unit_test_result.get_unit_test_result()

## @}
# @}

File: test/unit_test/src/unit_test_iec60730_bist.c
/***************************************************************************//**
 * @file  unit_test_iec60730_bist.c
 * @brief Unit test function sl_iec60730_bist().
 *******************************************************************************
 * # License
 * <b>Copyright 2020 Silicon Laboratories Inc. www.silabs.com</b>
 *******************************************************************************
 *
 * The licensor of this software is Silicon Laboratories Inc. Your use of this
 * software is governed by the terms of Silicon Labs Master Software License
 * Agreement (MSLA) available at
 * www.silabs.com/about-us/legal/master-software-license-agreement. This
 * software is distributed to you in Source Code format and is governed by the
 * sections of the MSLA applicable to Source Code.
 *
 ******************************************************************************/

#include "unit_test_common.h"
#include "unit_test_iec60730_bist.h"
#include "sl_iec60730_internal.h"

/*=======Mock Code=====*/
static bool is_function_called = false;

__WEAK void sl_iec60730_safe_state(sl_iec60730_test_failure_t failure){
  (void)failure;
  is_function_called = true;
}

// Dump functions

__WEAK sl_iec60730_test_result_t sl_iec60730_vmc_bist(void)
{
  return cmock_return_value();
}

__WEAK sl_iec60730_test_result_t sl_iec60730_imc_bist(void)
{
  return cmock_return_value();
}

__WEAK sl_iec60730_test_result_t sl_iec60730_cpu_registers_bist(void)
{
  return cmock_return_value();
}

__WEAK void sl_iec60730_restart_watchdogs(void) {

}

/*=======Test Case=====*/
void test_sl_iec60730_bist_pass_all_check_condition(void)
{
  /*Setup*/
  is_function_called = false;
  cmock_set_value(IEC60730_TEST_PASSED);
  /*Execute test*/
  sl_iec60730_bist();
  TEST_ASSERT_EQUAL(false, is_function_called);
}

void test_sl_iec60730_bist_failed_check_condition(void)
{
  /*Setup*/
  is_function_called = false;
  cmock_set_value(IEC60730_TEST_FAILED);
  /*Execute test*/
  sl_iec60730_bist();
  TEST_ASSERT_EQUAL(true, is_function_called);
}

void test_iec60730_safety_check_error_occur(void)
{
  /*Setup*/
  is_function_called = false;
  sl_iec60730_safety_check_error_occur(IEC60730_CLOCK_FAIL);
  cmock_set_value(IEC60730_TEST_PASSED);
  /*Execute test*/
  sl_iec60730_bist();
  TEST_ASSERT_EQUAL(true, is_function_called);
}

/*=======Run Test Case=====*/

void unit_test_run_all_test_cases(void)
{
  // Initialize
  memset((void *)test_case_result, 0xff, sizeof(test_case_result));
  current_test_numth = 0;

  // Start run test
  printf("--- BEGIN UNIT TEST ---\n");
  UnityBegin("/unit_test/src/unit_test_iec60730_bist.c");
  run_test(test_sl_iec60730_bist_pass_all_check_condition,"test_sl_iec60730_bist_pass_all_check_condition",54);
  run_test(test_sl_iec60730_bist_failed_check_condition,"test_sl_iec60730_bist_failed_check_condition",64);
  run_test(test_iec60730_safety_check_error_occur,"test_iec60730_safety_check_error_occur",74);
  UnityEnd();
  __asm volatile("IEC60730_UNIT_TEST_END:");

  while(1){
    // Do nothing
  }
}